From 5b13c44b871c966bedf5f34dc6e863e65652cb2b Mon Sep 17 00:00:00 2001
From: mdecimus <mauro@stalw.art>
Date: Thu, 26 Jun 2025 15:47:33 +0200
Subject: [PATCH] Fix OSS build (closes #1724)

---
 crates/common/src/core.rs | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/crates/common/src/core.rs b/crates/common/src/core.rs
index 3f92bd4c..0d6ff6b7 100644
--- a/crates/common/src/core.rs
+++ b/crates/common/src/core.rs
@@ -752,6 +752,14 @@ impl Server {
             .await
             .caused_by(trc::location!())
     }
+
+    #[cfg(not(feature = "enterprise"))]
+    pub async fn logo_resource(
+        &self,
+        _: &str,
+    ) -> trc::Result<Option<crate::manager::webadmin::Resource<Vec<u8>>>> {
+        Ok(None)
+    }
 }
 
 pub trait BuildServer {
-- 
2.49.0

